src/test/assembly/x86_64-simd-mask-reduce.rs

```diff
@@ -0,0 +1,33 @@
+// verify that simd mask reductions do not introduce additional bit shift operations
+
+// assembly-output: emit-asm
+// compile-flags: --crate-type=lib -O -C llvm-args=-x86-asm-syntax=intel
+// only-x86_64
```

Comment:
The tests could be deduplicated slightly if you made these tests `#[no_core]` and to use revisions and `--target=$TARGET` (grep for `revisions` in the test suite for usage examples).

These tests look like they wouldn’t be too hard to adapt to `#[no_core]` either, given that `Copy` and `Clone` derives are superfluous…

Reply:
I'm looking into combining the tests, but so far I could not get the `CHECK` annotations to work, or rather now they only work for one revision. Looking at [runtest.rs](https://github.com/rust-lang/rust/blob/31f858d9a511f24fedb8ed997b28304fec809630/src/tools/compiletest/src/runtest.rs#L2311) I think the following should work:

```rust
// revisions: aarch64 x86_64
// [x86_64] compile-flags: --target=x86_64-unknown-linux-gnu -C llvm-args=-x86-asm-syntax=intel
// [x86_64] needs-llvm-components: x86_64
// [aarch64] compile-flags: --target=aarch64-unknown-linux-gnu
// [aarch64] needs-llvm-components: aarch64
// assembly-output: emit-asm
// compile-flags: --crate-type=lib -O

...

pub unsafe fn mask_reduce_all(m: mask8x16) -> bool {
    // CHECK,NONMSVC,x86_64-NOT: psllw
    // CHECK,NONMSVC,x86_64: pmovmskb
    // CHECK,NONMSVC,x86_64: foo
    // CHECK,NONMSVC,aarch64: umaxv
    // CHECK,NONMSVC,aarch64: foo
    simd_reduce_all(m)
}
```

The `aarch64` checks seem to work, since I can cause them to fail by looking for `foo`. The `x86_64` revision always passes though.

Reply:
Writing `CHECK`s for specific revisions is just replacing `CHECK` with the name of the revision. So

```rust
    // x86_64-NOT: psllw
    // x86_64: pmovmskb
    // x86_64: foo
    // aarch64: umaxv
    // aarch64: foo
```

If you wanted to check non-msvc separately, you would need a separate revision (though isn't that already checked by the fact that both revisions are compiling for a `-gnu` target?)

I’m sorry we don't have good docs for this T_T

Reply:
Thanks, I'll take another look at this, my previous assumption was that the whole prefix was matched as a single string.